Add attr def for position, direction, angle and gradient support

fxbt opened the pull request.

ezyang commented on library/HTMLPurifier/AttrDef/CSS/Gradient.php:
```diff
@@ -0,0 +1,153 @@
+
+/**
+ * Validates Gradient as defined by CSS.
+ * Works with
```
Truncated!

ezyang commented on library/HTMLPurifier/AttrDef/CSS/Gradient.php:
```diff
@@ -0,0 +1,153 @@
+
+        // Get gradient function (linear, radial, repeating-linear, repeating-radial)
+        $function = explode('gradient', $string);
+        $function = reset($function) . 'gradient';
```
I find this line exceptionally strange. Why are you concatenating gradient at the end of function?

ezyang commented on library/HTMLPurifier/AttrDef/CSS/Gradient.php:
```diff
@@ -0,0 +1,153 @@
+
+            // test for color
+            $size = false;
+            if (preg_match('#(.*)\s+(\d+(\.\d+)?(%|px|ex|em|in|cm|mm|pt|pc))\s*$#', $part, $matches)) {
```
Ugh, repeating the units of measure here is bad juju.

ezyang commented on library/HTMLPurifier/AttrDef/CSS/Gradient.php:
```diff
@@ -0,0 +1,153 @@
+            return false;
+        }
+
+        return $function . '(' . rtrim($final, ', ') . ')';
```
It doesn't look like the value of function was ever validated. There is only a fixed number of allowed ones, right?

fxbt replied:
You're right. I added a validation at the start of the method

ezyang commented on library/HTMLPurifier/AttrDef/CSS/Gradient.php:
```diff
@@ -0,0 +1,153 @@
+            }
+
+            if ($i === 1 && $linear === false) {
+                if (in_array(trim($part), ['circle', 'ellipse'], true)) {
```
Ick. Do the trim once, and make a little hash table please?

fxbt commented:
Hi @ezyang 

Thanks for your comments and sorry for the time it took me to read them. I changed a few things based on your review.

I added some validation, some comments on lines that weren't easy to understand and i tried to be compatible with PHP 5.3 (i didn't test it though).

Let me know if i have to change something else. 😄

ezyang commented:
Looks like there is some sort of PHP 5.3 specific memory usage problem CI is picking up on. I don't have time to debug, maybe you can take a look.

fxbt commented:
I was not able to get the memory usage problem on my dev environment. Looks like simpletest got too many instances in memory and it fails.

To fix this, i increased the memory limit to 128M (64M before) but i don't know if this is the right solution.

fxbt commented:
Do I need to do something else on this pull request ?

ezyang commented:
I have to find some time to review the code carefully, sorry about the delay!

beebop1032 commented:
@ezyang This pull request corrects a problem I have in my project, could you integrate it?

ezyang commented on library/HTMLPurifier/AttrDef/CSS/Angle.php:
```diff
@@ -0,0 +1,29 @@
+            return false;
+        }
+
+        $value = (float)($matches[1] . $matches[2]);
```
Converting into a float will cause some weird precision problems. Since your regex is pretty strict already this conversion seems unnecessary.

ezyang commented on library/HTMLPurifier/AttrDef/CSS/Gradient.php:
```diff
@@ -0,0 +1,171 @@
+            $linear = true;
+        }
+
+        preg_match('#gradient\((.*)\)#', $string, $values);
```
This regex will get bamboozled if I have something like `repeating-linear-gradient-gradient()`, won't it?

ezyang commented:
@beebop1032 I'm probably not going to merge this patchset as is (too much fiddly string manipulation makes it hard to review) but it looks reasonable, so you should feel free to pull it and use it in your project.
